### PR TITLE
Improve screen-reader experience

### PR DESCRIPTION
**Pull request by jonalport**
https://govausites.atlassian.net/browse/SC-136

Added behaviour includes:
* Navigating between cities lands you on the overview page (as opposed to retaining subnav context) and gives focus to overview subnav item
* Navigating to subnav items gives focus to the heading pertaining to that content
* When focus is shifted, the visual outline style is removed (for that one focus only) to avoid distraction
* For links with hashes, focus is given to the element with that hash/id (Eg. for additional info pages)
* Aria labels on subnav buttons and subnav content headings, specifying which city that link/heading applies to (Eg. "Overview of Canberra", "Housing in Sydney")
* Renamed the ScrollTo... component to be more generic, as it now handles the focus stuff too

### CODE REVIEW

**Comment by jonalport** [comment]
This also appears bigger than it is. I probably should have `git moved` those renamed files...

File: package.json
Comment (davidgilbertson) [inline package.json]: You shouldn't need this, you can just hit `u` when running tests to update the snapshots.

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (davidgilbertson) [inline src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: a) a less whimsical name would probably be a good idea, maybe `RouteChangeHandler`.
b) if you rename the file in a separate branch (call it tech debt) then your changes to this file will show as changes, not a whole new file. As is, all history will be lost. Not a big deal in this case.

(FYI git won't track renamed files directly, not even with `git mv`)

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (davidgilbertson) [inline src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: This could be simplified to `focusPoint.style.outline = '';` and the `outlineStyle` variable removed. Or if you wanted to get fancy, `focusPoint.style.removeProperty('outline');`.

Setting a style property to an empty string calls `removeProperty()` under the hood anyway.

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (davidgilbertson) [inline src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: You don't need to wait until the element is focused to do this, just hide the outline and set tab index immediately, and remove them again on blur:

```js
  focusPoint.setAttribute('tabIndex', '0');
  focusPoint.style.outline = 'none';

  addOneTimeEventListener(focusPoint, 'blur', () => {
    focusPoint.removeAttribute('tabIndex');
    focusPoint.style.outline = '';
  });

  focusPoint.focus();
```

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: For consistency, prefer one per line.

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: I think it's better to not destructure props like this, and I should probably have that in the readme.

It makes it more difficult to discern what values are coming from props and which are declared locally, this is more apparent in bigger components where there might be half a dozen each of local variables and props.

(And just for consistency, everywhere else props are referred to by `props.xxx`).

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: I think `tabIndex` should be added programatically at the same point you're setting focus, otherwise this is always in the tab order, and can be highlighted:

image

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: I don't think this is necessary. Doesn't the text for the navigation link read "Infrastructure for Bendigo"? So when they click on that we don't need to repeat the city. Maybe ask Lily for a second opinion. I'd err on the side of less complexity.

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: You don't need to export this. The fact that it's a separate function is just an implementation detail of the component. So when testing the component you can just check the `aria-label` property. If it's right, then this function is right.

(But also, I don't think it's needed at all here, see below).

File: src/components/AdditionalInfoPage/AdditionalInfoPage.js
Comment (davidgilbertson) [inline src/components/AdditionalInfoPage/AdditionalInfoPage.js]: These should only be temporarily focusable. This leaves the `<h1>` in the tab order (see below).

File: src/components/CategoryNavContainer/CategoryNavContainer.js
Comment (davidgilbertson) [inline src/components/CategoryNavContainer/CategoryNavContainer.js]: You could probably simplify this and just pass the city if there is one, e.g. 

```js
city: state.cities.find(city => city.id === ownProps.cityId),
```

And then in the nav handle the case of 'all cities'.

File: src/components/CategoryNavContainer/CategoryNavContainer.js
Comment (davidgilbertson) [inline src/components/CategoryNavContainer/CategoryNavContainer.js]: Same comment as elsewhere. Rather than export implementation details for direct testing, just test that the container returns what it should.

File: src/components/FeedbackPage/FeedbackPage.js
Comment (davidgilbertson) [inline src/components/FeedbackPage/FeedbackPage.js]: All these `tabIndex="0"` can go

File: src/constants/misc.js
Comment (davidgilbertson) [inline src/constants/misc.js]: A bit further down in this file we have a `STRINGS` bucket for all strings.

File: src/components/AllCitiesOverviewContainer/AllCitiesOverview/AllCitiesOverview.js
Comment (davidgilbertson) [inline src/components/AllCitiesOverviewContainer/AllCitiesOverview/AllCitiesOverview.js:49]: Love a good descriptive variable name.

**Review by davidgilbertson — CHANGES_REQUESTED** [review]
A few things...

**Comment by jonalport** [comment]
Thanks for the detailed review, @davidgilbertson! Am currently addressing.

File: package.json
Comment (jonalport) [reply package.json]: Cool, that's easier.

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (jonalport) [reply src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: Agreed, but how do you know it will be ''? The idea behind saving the initial value is that it restores it to whatever it was.

**Comment by jonalport** [comment]
Test comment...

File: src/components/PageBanner/PageBanner.js
Comment (jonalport) [reply src/components/PageBanner/PageBanner.js]: Yeah I can see that argument, but it also makes things more concise not to have "this.props." 50 times in a file. Will take it out for consistency though 👍

File: src/components/PageBanner/PageBanner.js
Comment (jonalport) [reply src/components/PageBanner/PageBanner.js]: Yes, this is a better idea.

File: src/components/PageBanner/PageBanner.js
Comment (jonalport) [reply src/components/PageBanner/PageBanner.js]: Complexity vs consistency... Have taken it out for headings 👍

File: src/constants/misc.js
Comment (jonalport) [reply src/constants/misc.js]: Roger that.

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (jonalport) [reply src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: But where's the fun in that 😁 

I guess I assumed git mv would have eased the diffing overload but perhaps not then 😞

File: src/components/CategoryNavContainer/CategoryNavContainer.js
Comment (jonalport) [reply src/components/CategoryNavContainer/CategoryNavContainer.js]: Yep I will just pass the name down along with the ID then.

File: src/components/CategoryNavContainer/CategoryNavContainer.js
Comment (jonalport) [reply src/components/CategoryNavContainer/CategoryNavContainer.js]: Yeah I do agree with this. Black box!

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (davidgilbertson) [reply src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: Yeah I thought of that after making the comment.

File: src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js
Comment (davidgilbertson) [reply src/components/DoStuffOnRouteChange/DoStuffOnRouteChange.js]: Yep, they probably should never have made a `git mv` - it's just confusing. Bottom of [this page](https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository) shows that it just does a normal move.

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [reply src/components/PageBanner/PageBanner.js]: You can still do `const { props } = this`.

We actually had a bug in the ICT project where none of the props are referenced with `prop.x` and someone was doing the equivalent of `cities.sort()`, not realising they were mutating state and altering sorting elsewhere.

**Comment by jonalport** [comment]
I think I've now addressed all the things in the review. Is there a workflow step that I need to complete? Or @davidgilbertson do you need to confirm (to unblock)? 

😕

File: src/components/AdditionalInfoPage/AdditionalInfoPage.js
Comment (davidgilbertson) [inline src/components/AdditionalInfoPage/AdditionalInfoPage.js]: This is adding new blank lines, there should be no changes in this file

File: src/components/CategoryNavContainer/CategoryNav/CategoryNav.js
Comment (davidgilbertson) [inline src/components/CategoryNavContainer/CategoryNav/CategoryNav.js]: Should this be `.isRequired`?

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: Remove commented-out code

File: src/components/PageBanner/PageBanner.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.js]: Extra blank line can go.

File: src/components/PageBanner/PageBanner.test.js
Comment (davidgilbertson) [inline src/components/PageBanner/PageBanner.test.js]: This is never used (and doesn't exist), can go.

File: src/components/RouteChangeHandler/RouteChangeHandler.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.js]: This needs to `focusPoint.removeAttribute('tabIndex')` otherwise:

image

File: src/components/RouteChangeHandler/RouteChangeHandler.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.js]: A very minor, opinion thing, that isn't new in these latest changes: this might be easier to read with named variables, e.g.

```js
const oldCity = oldPath.split('/')[1];
const [, currentCity, currentCategory] = currentPath.split('/');

if (currentCategory === NO_CATEGORY && currentCity !== oldCity) {
  focusPoint = document.getElementById(ELEMENT_IDS.OVERVIEW_LINK);
}
```

File: src/components/RouteChangeHandler/RouteChangeHandler.test.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.test.js]: Since every discovered bug should have a test, this should be testing that `tabIndex` and `outline` are reset on blur.

File: src/components/RouteChangeHandler/RouteChangeHandler.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.js]: Minor (and not new in these change) - this is logically just an `else`, right? Might be easier to read like so:
```js
// first look for a hash
if (document.location.hash) {
  focusPoint = document.querySelector(document.location.hash);
} else {
  // if we have changed cities, and landed on overview
  // then focus on overview subnav item
  const oldSegments = oldPath.split('/', 3);
  const currentSegments = currentPath.split('/', 3);

  if (currentSegments[2] === NO_CATEGORY && currentSegments[1] !== oldSegments[1]) {
    focusPoint = document.getElementById(ELEMENT_IDS.OVERVIEW_LINK);
  }
}
```

**Review by davidgilbertson — CHANGES_REQUESTED** [review]
A few issues

File: src/components/RouteChangeHandler/RouteChangeHandler.js
Comment (jonalport) [reply src/components/RouteChangeHandler/RouteChangeHandler.js]: DOH!

File: src/components/RouteChangeHandler/RouteChangeHandler.test.js
Comment (jonalport) [reply src/components/RouteChangeHandler/RouteChangeHandler.test.js]: Roger that

File: src/components/RouteChangeHandler/RouteChangeHandler.js
Comment (jonalport) [reply src/components/RouteChangeHandler/RouteChangeHandler.js]: Array destructure, smart. Definitely reads better.

File: src/components/AdditionalInfoPage/AdditionalInfoPage.js
Comment (jonalport) [reply src/components/AdditionalInfoPage/AdditionalInfoPage.js]: Ah yes, that was my find and replace, whoops. We should make the linter pick this up.

File: .eslintrc.js
Comment (davidgilbertson) [inline .eslintrc.js]: This should be reverted now since we're not breaking the rule any more.

File: src/components/CategoryNavContainer/CategoryNav/CategoryNav.js
Comment (davidgilbertson) [inline src/components/CategoryNavContainer/CategoryNav/CategoryNav.js:91]: This throws a propTypes error in the console on all cities/overview. I thought it was always passed in. So you could either put ` || STRINGS.ALL_CITIES` somewhere further up the tree or revert to not required since you have `const cityName = this.props.cityName || STRINGS.ALL_CITIES;` in this file already.

Although the preferred approach for logic like `const cityName = this.props.cityName || STRINGS.ALL_CITIES;` is to use defaultProps and keep the component cleaner. That's a minor thing though.

File: src/components/CategoryNavContainer/CategoryNav/CategoryNav.test.js
Comment (davidgilbertson) [inline src/components/CategoryNavContainer/CategoryNav/CategoryNav.test.js]: If you revert to not having `isRequired` this can go.

File: src/components/RouteChangeHandler/RouteChangeHandler.test.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.test.js:162]: Nice.

File: src/components/RouteChangeHandler/RouteChangeHandler.test.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.test.js]: Very minor, redundant variable, e.g. could be
```js
const domNode = wrapper.find(`#${focusElementId}`).getDOMNode();
```

File: src/components/RouteChangeHandler/RouteChangeHandler.test.js
Comment (davidgilbertson) [inline src/components/RouteChangeHandler/RouteChangeHandler.test.js]: Typo? "a it tries"

**Review by davidgilbertson — CHANGES_REQUESTED** [review]
propTypes error and some tidy up stuff

File: src/constants/misc.js
Comment (jonalport) [inline src/constants/misc.js:5]: If they are all 14, this will still be a problem. Better than nothing though.

File: src/constants/categories.test.js
Comment (davidgilbertson) [inline src/constants/categories.test.js:69]: Minor. Really I just want to make one more comment...

Technically `const name = category.navName || category.name;` could go up two lines and be used inside the `it()` as well.

**Review by davidgilbertson — APPROVED** [review]
LGTM!

File: src/constants/categories.test.js
Comment (jonalport) [reply src/constants/categories.test.js:69]: Yes, this is true. I wanted to package it neatly together.

**Pull request merged**
